sdks/java/harness/src/main/java/org/apache/beam/fn/harness/WindowMergingFnRunner.java

```diff
@@ -151,10 +151,12 @@ public void merge(Collection<W> toBeMerged, W mergeResult) throws Exception {
         KV<T, Iterable<W>> windowsToMerge) throws Exception {
       currentWindows = Sets.newHashSet(windowsToMerge.getValue());
       windowFn.mergeWindows((MergeContext) mergeContext);
+      ArrayList mergedWindowsCopy = new ArrayList<>(mergedWindows);
       for (KV<W, Collection<W>> mergedWindow : mergedWindows) {
         currentWindows.removeAll(mergedWindow.getValue());
       }
-      return KV.of(windowsToMerge.getKey(), KV.of(currentWindows, (Iterable) mergedWindows));
+      mergedWindows.clear();
```

Comment:
It seems like it's a bug which can result in wrong results before this change, right? If it's not too complicated, can we add a unit test like FnApiDoFnRunnerTest to guard this cleanup logic?

Reply:
It definitely can be considered wrong, since the previous merge result could have been processed by runner and the related windows could have been cleaned up. Added in unit test.